### PR TITLE
Add support for batched pdist

### LINKED ISSUE #9406: [pytorch] [feature request] Pairwise distances between all points in a set (a true pdist)
Currently [F.pairwise_distance](https://pytorch.org/docs/stable/nn.html?highlight=pairwise#torch.nn.functional.pairwise_distance) and [F.cosine_similarity](https://pytorch.org/docs/stable/nn.html?highlight=pairwise#torch.nn.functional.cosine_similarity) accept two sets of vectors of the same size and compute similarity between corresponding vectors.

However, it's often useful to compute pairwise similarities or distances between all points of the set (in mini-batch metric learning scenarios), or between all possible pairs of two sets (e.g. to build a bi-partite weighted graph). It's trivial to write ad-hoc functions like this, but I believe it's also a useful and a frequent primitive (for euclidean, square euclidean, etc).

One way to introduce this would be modifying `pairwise_distance` to allow other shapes and an argument for metric type.

Such a function for example exists in SciPy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html (computes all pairwise distances within one set of vectors)

Also exists in MatLab: https://www.mathworks.com/help/stats/pdist.html

### PR DESCRIPTION
**Pull request by erikbrinkman**
Summary: This updates pdist to work for batched inputs, and updates the
documentation to reflect issues raised.

closes #9406

Test Plan: Batched versions were added to old tests, and the old scipy
test switched to using the internal norm calculation, which also
illustrates alternate ways to compute pdist with norm.

### CODE REVIEW

File: aten/src/ATen/native/Distance.cpp [hunk 1]
```diff
@@ -14,9 +14,12 @@ Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double ep
 }
 
 // This is to guarantee that the contiguous memory is passed to the backward pass
+// TODO This currently enforces that the entire array is contiguous, but the
+// batches don't really have to be for efficiency sake, meaning there may be a
+// better way to only force the last two dimensions as contiguous.
```
Comment (ezyang) [inline aten/src/ATen/native/Distance.cpp]: It's not too hard to do a contiguity check for just the last two dims; you just need to check that stride[d-1] is 1, and stride[d-2] is size[d-1].

File: aten/src/ATen/native/Distance.cpp [hunk 2]
```diff
@@ -26,17 +29,23 @@ Tensor _pdist_forward(const Tensor& self, const double p) {
...
-    if (self.size(1) == 0) {
+
+  const auto batches = self.sizes().slice(0, self.dim() - 2);
+  int64_t b = at::tensor(batches).prod().item<int64_t>();
+  int64_t n = self.size(-2);
+  int64_t m = self.size(-1);
+  int64_t c = n * (n - 1) / 2;
+
+  std::vector<int64_t> result_sizes(batches.begin(), batches.end());
+  result_sizes.push_back(c);
```
Comment (ezyang) [inline aten/src/ATen/native/Distance.cpp]: nit: you are almost certainly triggering a realloc here

**Review by ezyang — APPROVED** [review]
okay-dokey

Do we have some perf numbers?

File: aten/src/ATen/native/Distance.cpp (hunk 1 above)
Comment (erikbrinkman) [reply aten/src/ATen/native/Distance.cpp]: Yeah, but then I'd also have to work on more general handling of the remaining dimensions. I think there's probably a more general way to write this code, I just wanted to make sure that some form of pdist is available for batched operations. Also, I thinks it's very unlikely for this whole tensor to not be contiguous, so this really shouldn't be that big of an issue.

**Comment by gpleiss** [comment]
Is there a plan for merging this PR? This would really improve some memory/speed issues we've had with GPyTorch (https://github.com/cornellius-gp/gpytorch/issues/438)

**Comment by ezyang** [comment]
This diff is blocked on #15406
